File: services/cognitive-state-engine/src/nova_cognitive_state_engine/api/cognitive_state.py
"""`/v1/cognitive-state/*` -- TDD 4F.7 §7's three routes, and nothing beyond them.

| Method | Path |
|---|---|
| `GET` | `/v1/cognitive-state/thoughts` |
| `GET` | `/v1/cognitive-state/focus` |
| `GET` | `/v1/cognitive-state/sensors` |

**Read-only, by construction** (RS-1b). Only `@router.get` is declared, no route
has a body, a path parameter or a query parameter, and nothing here can reach
`upsert_thought`, `move_layer` or `promote_thought`. Any other method answers
405, which `api-gateway` forwards unchanged. Creating thoughts, authoring
`ProposedAction`s and promoting are the promotion slice 4F.P's (RS-1c).

**Identity is the server's** (ADR-025, RS-4a). Every thought read resolves
`Settings.primary_user_id`; there is no `user_id` parameter, so a caller cannot
name another identity -- a `?user_id=` is simply ignored.

**Empty is empty.** An empty store answers `{"thoughts": []}`, `{"capacity": …,
"entries": []}` and `{"sensors": []}`. Nothing is synthesized. **An error is
never an empty success**: a store or validation failure propagates as an error
response, because `[]` is a claim that nothing exists.

**No autonomy data** (RS-4b). No route reads `autonomy-engine`, and no response
carries a decision, outcome, `subject_id` or triggered/executed field.
`proposed_action` is returned **as a proposal** (RS-4c) -- the persisted value,
exactly, and nothing about what happened to it.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Request
from pydantic import BaseModel, ConfigDict

from nova_cognitive_state_engine.domain.focus import select_focus
from nova_cognitive_state_engine.domain.models import (
    ActiveThought,
    FocusedThought,
    ProposedAction,
)
from nova_cognitive_state_engine.domain.sensor_state import (
    SensorLifecycleState,
    SensorStateRecord,
)

__all__ = [
    "FocusEntryResponse",
    "FocusResponse",
    "ProposedActionResponse",
    "SensorStateListResponse",
    "SensorStateResponse",
    "ThoughtListResponse",
    "ThoughtResponse",
    "router",
]

router = APIRouter(prefix="/v1/cognitive-state", tags=["cognitive-state"])


class _Strict(BaseModel):
    """Every 4F.7 response model forbids unknown fields -- 4D's and 4E's
    precedent (`digital-twin-engine/api/digital_twin.py`). A field that is not
    in TDD 4F.7 §7 cannot be added by accident."""

    model_config = ConfigDict(extra="forbid")


class ProposedActionResponse(_Strict):
    """`ProposedAction`, verbatim -- a **proposal**, and nothing about its fate."""

    category: str
    risk: str
    action_type: str
    execution_target: str
    verification_method: str
    title: str
    detail: str

    @classmethod
    def of(cls, action: ProposedAction) -> ProposedActionResponse:
        return cls(
            category=action.category.value,
            risk=action.risk.value,
            action_type=action.action_type,
            execution_target=action.execution_target,
            verification_method=action.verification_method,
            title=action.title,
            detail=action.detail,
        )


class ThoughtResponse(_Strict):
    """Every Part 6 field of an Active Thought (l.123–137), plus its identity,
    Attention Layer and bookkeeping. **No `user_id`**: identity is the
    server's, and echoing it gives a client nothing to act on."""

    thought_id: UUID
    description: str
    priority: int
    confidence: float
    dependencies: list[UUID]
    estimated_completion: datetime | None
    related_memories: list[UUID]
    related_projects: list[UUID]
    current_progress: float
    attention_layer: Literal["immediate", "active", "passive", "dormant", "archived"]
    created_at: datetime
    updated_at: datetime
    proposed_action: ProposedActionResponse | None

    @classmethod
    def of(cls, thought: ActiveThought) -> ThoughtResponse:
        return cls(
            thought_id=thought.thought_id,
            description=thought.description,
            priority=thought.priority,
            confidence=thought.confidence,
            dependencies=list(thought.dependencies),
            estimated_completion=thought.estimated_completion,
            related_memories=list(thought.related_memories),
            related_projects=list(thought.related_projects),
            current_progress=thought.current_progress,
            attention_layer=thought.attention_layer.value,
            created_at=thought.created_at,
            updated_at=thought.updated_at,
            proposed_action=(
                ProposedActionResponse.of(thought.proposed_action)
                if thought.proposed_action is not None
                else None
            ),
        )


class ThoughtListResponse(_Strict):
    thoughts: list[ThoughtResponse]


class FocusEntryResponse(_Strict):
    """One `FocusedThought`. `signals_used` is returned even when empty --
    which, with no focus signal computed anywhere in production, is always."""

    thought: ThoughtResponse
    score: float
    signals_used: list[str]

    @classmethod
    def of(cls, entry: FocusedThought) -> FocusEntryResponse:
        return cls(
            thought=ThoughtResponse.of(entry.thought),
            score=entry.score,
            signals_used=[signal.value for signal in entry.signals_used],
        )


class FocusResponse(_Strict):
    capacity: int
    entries: list[FocusEntryResponse]


class SensorStateResponse(_Strict):
    """One current record from `cognitive_state.sensor_state` (A-4F7-1).

    `state` is `perception-engine`'s `SensorState` vocabulary exactly (A-4F7-2),
    under the same field name `perception-engine`'s own REST uses. `reported_at`
    is dispatch time, not transition time (TDD 4F.7 §7.3)."""

    sensor_id: str
    sensor_type: str
    state: SensorLifecycleState
    reported_at: datetime

    @classmethod
    def of(cls, record: SensorStateRecord) -> SensorStateResponse:
        return cls(
            sensor_id=record.sensor_id,
            sensor_type=record.sensor_type,
            state=record.state,
            reported_at=record.reported_at,
        )


class SensorStateListResponse(_Strict):
    sensors: list[SensorStateResponse]


@router.get("/thoughts", response_model=ThoughtListResponse)
async def list_active_thoughts(request: Request) -> ThoughtListResponse:
    state = request.app.state
    thoughts = await state.repository.list_thoughts(user_id=state.settings.primary_user_id)
    return ThoughtListResponse(thoughts=[ThoughtResponse.of(thought) for thought in thoughts])


@router.get("/focus", response_model=FocusResponse)
async def current_focus(request: Request) -> FocusResponse:
    """`select_focus`, unchanged, with **no** signals supplied: focus-signal
    computation is OPEN, and supplying one here would decide it."""
    state = request.app.state
    capacity = state.settings.focus_capacity
    thoughts = await state.repository.list_thoughts(user_id=state.settings.primary_user_id)
    entries = select_focus(thoughts, capacity=capacity, inputs=None)
    return FocusResponse(
        capacity=capacity, entries=[FocusEntryResponse.of(entry) for entry in entries]
    )


@router.get("/sensors", response_model=SensorStateListResponse)
async def sensor_states(request: Request) -> SensorStateListResponse:
    records = await request.app.state.repository.list_sensor_states()
    return SensorStateListResponse(sensors=[SensorStateResponse.of(record) for record in records])

File: apps/web-client/src/entities/cognitiveState.ts
import { useQuery, useQueryClient } from "@tanstack/react-query";
import { z } from "zod";

import { gatewayFetch } from "./http";

/**
 * What NOVA is currently thinking about, and what it last heard from its
 * sensors -- `cognitive-state-engine`'s `/v1/cognitive-state/*` surface (Bible
 * Part 6; TDD 4F.7 §7), read through `api-gateway`.
 *
 * **Read-only, by construction.** Three `GET`s and nothing else: the engine
 * declares no other method (405 for every one), and this module declares no
 * mutation. The panel never creates or promotes a thought, never authors a
 * proposal, never asks for a decision, never executes anything, never changes
 * an autonomy level and never touches a sensor (RS-1b).
 *
 * **No polling and no realtime.** No `refetchInterval`, no `setInterval`, no
 * socket subscription (A-4F7-3 (a); D-4F-6). Data is read on mount, and again
 * only when the operator presses Refresh, which re-issues the same three
 * `GET`s. `perception.sensor.health_changed` does reach the browser on the
 * public topic, but the panel does not listen to it: a refetch triggered by
 * that frame would race the engine's own write and could show the old state.
 *
 * **No user identity is sent.** ADR-025 gives one trusted user per instance and
 * the engine resolves it server-side, so these paths carry no `user_id`.
 *
 * **The schemas are strict** (`.strict()`): an unknown field fails the parse
 * rather than being dropped, so a contract drift surfaces through the
 * degradation notice instead of as data the panel half-understands. The
 * sensor `state` is exactly `perception-engine`'s six lifecycle values
 * (A-4F7-2); any other value is a contract violation, never a state.
 */

/** Bible Part 6's five attention layers, in Part 6's order. */
export const ATTENTION_LAYERS = ["immediate", "active", "passive", "dormant", "archived"] as const;

/** `perception-engine`'s `SensorState`, exactly (A-4F7-2). No health vocabulary. */
export const SENSOR_STATES = [
  "uninitialized",
  "initialized",
  "running",
  "paused",
  "stopped",
  "failed",
] as const;

const proposedActionSchema = z
  .object({
    category: z.string(),
    risk: z.string(),
    action_type: z.string(),
    execution_target: z.string(),
    verification_method: z.string(),
    title: z.string(),
    detail: z.string(),
  })
  .strict();

const thoughtSchema = z
  .object({
    thought_id: z.string(),
    description: z.string(),
    priority: z.number(),
    confidence: z.number(),
    dependencies: z.array(z.string()),
    /** `null` is "not estimated" -- never a date, never a zero. */
    estimated_completion: z.string().nullable(),
    related_memories: z.array(z.string()),
    related_projects: z.array(z.string()),
    current_progress: z.number(),
    attention_layer: z.enum(ATTENTION_LAYERS),
    created_at: z.string(),
    updated_at: z.string(),
    /** A proposal NOVA holds, and nothing about what happened to it (RS-4c). */
    proposed_action: proposedActionSchema.nullable(),
  })
  .strict();

const thoughtListSchema = z.object({ thoughts: z.array(thoughtSchema) }).strict();

const focusEntrySchema = z
  .object({
    thought: thoughtSchema,
    score: z.number(),
    /** Always `[]` today: no production code computes a focus signal (K-5). */
    signals_used: z.array(z.string()),
  })
  .strict();

const focusSchema = z
  .object({
    capacity: z.number(),
    entries: z.array(focusEntrySchema),
  })
  .strict();

const sensorSchema = z
  .object({
    sensor_id: z.string(),
    sensor_type: z.string(),
    state: z.enum(SENSOR_STATES),
    /** When `perception-engine`'s outbox dispatched the report (K-3). */
    reported_at: z.string(),
  })
  .strict();

const sensorListSchema = z.object({ sensors: z.array(sensorSchema) }).strict();

export const cognitiveStateSchemas = {
  thoughts: thoughtListSchema,
  focus: focusSchema,
  sensors: sensorListSchema,
} as const;

export type AttentionLayer = (typeof ATTENTION_LAYERS)[number];
export type SensorLifecycleState = (typeof SENSOR_STATES)[number];
export type ThoughtView = z.infer<typeof thoughtSchema>;
export type ProposedActionView = z.infer<typeof proposedActionSchema>;
export type FocusView = z.infer<typeof focusSchema>;
export type FocusEntryView = z.infer<typeof focusEntrySchema>;
export type SensorView = z.infer<typeof sensorSchema>;

export const cognitiveStateKeys = {
  all: ["cognitive-state"] as const,
  thoughts: ["cognitive-state", "thoughts"] as const,
  focus: ["cognitive-state", "focus"] as const,
  sensors: ["cognitive-state", "sensors"] as const,
};

export function useThoughts() {
  return useQuery({
    queryKey: cognitiveStateKeys.thoughts,
    queryFn: async ({ signal }) =>
      await gatewayFetch("/v1/cognitive-state/thoughts", thoughtListSchema, { signal }),
  });
}

export function useFocus() {
  return useQuery({
    queryKey: cognitiveStateKeys.focus,
    queryFn: async ({ signal }) =>
      await gatewayFetch("/v1/cognitive-state/focus", focusSchema, { signal }),
  });
}

export function useSensors() {
  return useQuery({
    queryKey: cognitiveStateKeys.sensors,
    queryFn: async ({ signal }) =>
      await gatewayFetch("/v1/cognitive-state/sensors", sensorListSchema, { signal }),
  });
}

/**
 * **Refresh** -- re-issue the same three `GET`s, and nothing else.
 *
 * Not a mutation: nothing is written, nothing is patched into the cache, and
 * what renders afterwards is exactly what the engine answered. Invalidating
 * marks the three queries stale, and the mounted ones refetch through their
 * own `queryFn`s -- the same `GET`s the panel issued on mount.
 */
export function useRefreshCognitiveState(): () => Promise<void> {
  const queryClient = useQueryClient();
  return async () => {
    await queryClient.invalidateQueries({ queryKey: cognitiveStateKeys.all });
  };
}
